towhee/trainer/cnn_trainer.py

```diff
@@ -0,0 +1,629 @@
+import inspect
+import math
+import os
+import random
+import re
+import shutil
+import sys
+import time
+import warnings
+from logging import StreamHandler
+from pathlib import Path
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Tuple, Union
+
+from tqdm.auto import tqdm
+
+
+# Integrations must be imported before ML frameworks:
+from integrations import (  # isort: split
+    default_hp_search_backend,
+    get_reporting_integration_callbacks,
+    hp_params,
+    is_fairscale_available,
+    is_optuna_available,
+    is_ray_tune_available,
+    run_hp_search_optuna,
+    run_hp_search_ray,
+)
+
+import numpy as np
+import torch
+from packaging import version
+from torch import nn
+from torch.utils.data.dataloader import DataLoader
+from torch.utils.data.dataset import Dataset, IterableDataset
+
+from configuration_utils import PretrainedConfig
+from deepspeed import deepspeed_init, is_deepspeed_zero3_enabled
+from file_utils import (
+    CONFIG_NAME,
+    WEIGHTS_NAME,
+    is_datasets_available,
+)
+# from modelcard import TrainingSummary
+from modeling_utils import PreTrainedModel, unwrap_model
+from optimization import Adafactor, AdamW, get_scheduler
+from trainer_callback import (
+    CallbackHandler,
+    DefaultFlowCallback,
+    PrinterCallback,
+    ProgressCallback,
+    TrainerCallback,
+    TrainerControl,
+    TrainerState,
+)
+
+from trainer_pt_utils import (
+    DistributedLengthGroupedSampler,
+    DistributedSamplerWithLoop,
+    DistributedTensorGatherer,
+    IterableDatasetShard,
+    LabelSmoother,
+    LengthGroupedSampler,
+    SequentialDistributedSampler,
+    ShardSampler,
+    distributed_broadcast_scalars,
+    distributed_concat,
+    find_batch_size,
+    get_parameter_names,
+    nested_concat,
+    nested_detach,
+    nested_numpify,
+    nested_truncate,
+    nested_xla_mesh_reduce,
+    reissue_pt_warnings,
+)
+
+from trainer_utils import (
+    PREFIX_CHECKPOINT_DIR,
+    BestRun,
+    EvalLoopOutput,
+    EvalPrediction,
+    HPSearchBackend,
+    PredictionOutput,
+    ShardedDDPOption,
+    TrainerMemoryTracker,
+    TrainOutput,
+    default_compute_objective,
+    default_hp_space,
+    denumpify_detensorize,
+    get_last_checkpoint,
+    number_of_arguments,
+    set_seed,
+    speed_metrics,
+)
+
+from training_args import ParallelMode, TrainingArguments
+from utils import logging
+from utils.modeling_auto_mapping import MODEL_FOR_QUESTION_ANSWERING_MAPPING_NAMES
+
+
+_is_torch_generator_available = False
+_is_native_amp_available = False
+
+DEFAULT_CALLBACKS = [DefaultFlowCallback]
+DEFAULT_PROGRESS_CALLBACK = ProgressCallback
+
+
+if version.parse(torch.__version__) >= version.parse("1.6"):
+    _is_torch_generator_available = True
+    _is_native_amp_available = True
+    from torch.cuda.amp import autocast
+
+if is_datasets_available():
+    import datasets
+
+
+if TYPE_CHECKING:
+    import optuna
+
+logger = logging.get_logger(__name__)
+
+
+class Trainer:
```

Comment:
should we consider a more specific Trainer name such as TorchCNNTrainer, since we may have other types of Trainer. 
In addition, do we need a base class 'TorchTrainer' for those basic methods in Pytorch training?